lightning/src/chain/chaininterface.rs

```diff
@@ -207,25 +207,33 @@ impl ChainWatchedUtil {
 
 /// Utility for notifying listeners about new blocks, and handling block rescans if new watch
 /// data is registered.
-pub struct BlockNotifier {
-	listeners: Mutex<Vec<Weak<ChainListener>>>, //TODO(vmw): try removing Weak
+pub struct BlockNotifier<'a> {
+	ref_listeners: Mutex<Vec<&'a ChainListener>>,
+	arc_listeners: Mutex<Vec<Arc<ChainListener>>>,
```

Comment:
Yeah agreed, here's the two things I tried:
1. `listeners: Deref<Target = ChainListener>` aka the Deref's Target is a trait object. This didn't work because the compiler was defaults to static lifetimes for trait objs unless explicitly stated otherwise (which then excludes Arc's iiuc).
2. `listeners: Deref<M> where M::Target: ChainListener` aka a list of generics such that the concrete type is chosen at compile time. This didn't work because the list needs to be able to hold two different concrete types: a SimpleChannelMonitor and a ChannelManager. 
2a. So then I tried using an ChainListener enum that can hold either a SimpleChanMonitor or a ChanManager, so in theory the list could hold a monitor and a manager. But this didn't work because given this enum:

```
enum ChainListenerEnum {
    SimpleChanMonitor(SimpleChanMonitor)
    ChanManager(ChanManager)
}
```
you can't call `ChainListenerEnum::ChannelManager(channel_manager)` because the ChannelManager is then `move`'d and the compiler complains. 

Let me know this doesn't make sense, or if you have any other ideas to work around this hack!

Reply:
I was able to get the 1st to work (including with rust-lightning-bitcoinrpc building against net-tokio) at https://github.com/TheBlueMatt/rust-lightning/tree/2020-01-443-absurdism

Reply:
oh, cool! I still don't fully understand the solution but that's very cool. Squashed your changes into my commit.

Reply:
My *vague* understanding is that if you say Type : 'a it means that the Type doesn't need to exist after 'a...in this case if you say Deref<Target = B> : 'a then the whole thing (including B) doesn't need to exist after 'a.

Reply:
ah, i was interpreting `'a` as "it _can't_ exist after 'a".

Reply:
I mentally think "extends" but thats more of a Java-/OOP-ism.